exercises/practice/parallel-letter-frequency/.meta/example.cpp

```diff
@@ -0,0 +1,38 @@
+#include "parallel_letter_frequency.h"
+
+#include <algorithm>
+#include <cctype>
+#include <execution>
+
+namespace parallel_letter_frequency {
+
+namespace {
+
+[[nodiscard]] std::unordered_map<char, size_t> frequency(
+    std::string_view const text) {
+    std::unordered_map<char, size_t> freq;
+    for (unsigned char c : text) freq[std::tolower(c)] += 1;
+    return freq;
+}
+
+}  // namespace
+
+std::unordered_map<char, size_t> frequency(
+    std::vector<std::string_view> const& texts) {
+    std::vector<std::unordered_map<char, size_t>> freqs(texts.size());
+    // determine frequencies of texts in parallel
+    std::transform(std::execution::par_unseq, texts.cbegin(), texts.cend(),
+                   freqs.begin(), [](auto text) { return frequency(text); });
+
+    // combine individual frequencies (this happens sequentially)
+    std::unordered_map<char, size_t> totals{};
```

Comment:
If you want to go deeper into parallel, you could use reduce or do transform_reduce above.
Is there a reason against it that I don't see?

Reply:
`std::reduce` doesn't work on references, but always on values. So here, using `std::reduce` would require creating a new `std::unordered_map` in each step. That would create enough overhead to make it overall slower than just doing it sequentially. Honestly, using `reduce` here didn't even occur to me before your comment, probably because the way I do it now is also a common pattern when doing GPU programming: Parallelize what's easily parallelizable (determining letter frequency per text) and do the final reduce on the host (because we're dealing with comparatively few elements anyway). For a parallel reduce to make sense here, we'd need much more than ten texts. With thousands it could start to make sense, but maybe not even then. Either way, the reduce would need a different strategy than just using `std::reduce`. I can't think of anything that would justify the effort in the present case.